app/Actions/Album/Archive.php

```diff
@@ -178,8 +180,7 @@ private function compressAlbum(AbstractAlbum $album, array &$usedDirNames, ?stri
 				// in smart albums should be owned by the current user...
 				if (
 					($album instanceof BaseSmartAlbum || $album instanceof TagAlbum) &&
-					!AccessControl::is_current_user_or_admin($photo->owner_id) &&
-					!($photo->album_id === null ? $album->is_downloadable : $photo->album->is_downloadable)
+					!Gate::check(PhotoPolicy::DOWNLOAD, $photo)
```

Comment:
The new behavior is not identical to the old one in all cases. The reason is that the implementation of `PhotoPolicy::download()` takes a different decision path than what has been there before.

In particular the following case bothers me. Assume we have an unsorted photo (i.e. `$photo->album_id === null)  which happens to be part of a tag album and this tag album is going to be downloaded.

According the previous rules

```php
$photo->album_id === null ? $album->is_downloadable : $photo->album->is_downloadable
```

the decision whether the unsorted photo is downloadable or not was decided based on the tag album in lack of a regular parent album.

Now, the new Photo policy is

```php
public function download(?User $user, Photo $photo): bool {
	if (!$this->visible($user, $photo)) {
		return false;
	}

	return $this->albumPolicy->download($user, $photo->album);
}
```

We first check whether the photo is visible. We haven't done this before, but I guess it should not harm. (Maybe it becomes less efficient for many photos at once.) After that the decision is forwarded to the new album policy.

Here, we have the *first two differences*:

 1. The policy does not consider ownership of the photo as it has been done before. (`PhotoPolicy::before` only checks for admin, but not for ownership, too.)
 2. Previously, we considered the tag album which is going to be downloaded as a fallback substitute, now we are always using the parent album of the photo as the decisive element which is `null` in the given example.

The new album policy is

```php
public function download(?User $user, ?BaseAlbum $baseAlbum): bool {
	if ($baseAlbum === null) {
		return Configs::getValueAsBool('downloadable');
	}

	return $this->own($user, $baseAlbum) || $baseAlbum->is_downloadable;
	}
```

As the album is `null` (the photo is unsorted) the global configuration value decides.

I am not saying that the old or new behavior is better or worse. But it is definitely a change.

In particular, consider non-admin user. The non-admin user uploads unsorted photos and creates a tag album. (Note: The non-admin user is owner of all.) The non-admin user adds the unsorted photo to the tag album. With the old behavior the non-admin user could have downloaded the photo. With the new behavior a non-admin user can only download his own unsorted photos if enabled by configuration.

I would call this change in behavior surprising, because I assume users always expect to be able to do everything with photos (and album) owned by themselves.

@LycheeOrg/reviewers Any opinions?

This is the first file I am reviewing after the big refactoring and it has already taken me 1.5 hours to spot this difference. There are 122 files more to go. This being said, this review will take quite a while, if we don't want to risk surprising (and maybe unpleasant) changes in behavior.

I know that @ildyria is eager to get this PR merged before the next public release, but I am not that sure whether this is a clever idea. (At least not, if we want to publish the next public release in a timely manner.)

Maybe we should also wait until we have https://github.com/LycheeOrg/Lychee/pull/1414 in place. This is currently in a draft state and far from complete, but it aims at adding test cases like the one described above. This should allow us to spot these kind of changes more easily.

Reply:
> 1. The **policy does not consider ownership of the photo** as it has been done before. (PhotoPolicy::before only checks for admin, but not for ownership, too.)

That is incorrect. The ownership is checked in `visible`.
```php
	/**
	 * Defines whether the photo is visible to the current user.
	 *
	 * @param User|null $user
	 * @param Photo     $photo
	 *
	 * @return bool
	 */
	public function visible(?User $user, Photo $photo): bool
	{
		return $this->own($user, $photo) ||
		$photo->is_public ||
		$this->albumPolicy->access($user, $photo->album);
	}
```

Reply:
> That is incorrect. Yes before checks for admin. But the ownership is checked in visible

But this doesn't help here. `visible` is only called by `download` to exit early with `false` in the case that the photo is not even visible. The check for ownership in `visible` does not ensure such that `download` returns `true`.

Reply:
The owner must always be able to download their photos and albums, irrespective of _any_ settings; that's without question.

That said, I'm surprised by the "old" behavior. I'm pretty sure that's not how it used to be and I don't know when it changed and whether the change was intentional. It used to be like that:
* the owner can always download
* if not owner, then the *actual* parent album determines downloadability
* if no parent album, the global config determines it

Note that there are no smart or tag albums involved in the list above. They are not the actual parent albums and as such, in my opinion, they should not determine anything of that sort: it's super-counterintuitive if the same photo can't be downloaded when viewed from the *actual* parent album but it can when viewed through a tag album.

I guess the GUI currently allows users to change properties such as `is_downloadable` for tag albums. In my opinion, it shouldn't. In an ideal world, these properties shouldn't exist at all for tag albums.

Reply:
FYI, here's the relevant code in 4.4.0:
https://github.com/LycheeOrg/Lychee/blob/5b51765c7c66e6542ff2a6dcd3fb57393ff16717/app/Actions/Photo/Archive.php#L126-L135

Reply:
OK, I realized that I was looking at the wrong code. I should've quoted this:
https://github.com/LycheeOrg/Lychee/blob/5b51765c7c66e6542ff2a6dcd3fb57393ff16717/app/Actions/Album/Archive.php#L145-L154

So I agree with @nagmat84 that the old (4.4.0) behavior was different. However, it was different due to a bug. If you look at the comment in the quote above (which I wrote), it makes it quite clear that both for smart and tag albums it's supposed to consult the parent album. The second part of the comment appears obsolete, BTW, because nowadays we _can_ have somebody else's photos there, via `public_photos_hidden`.

There is one more thing though that bothers me about the quoted excerpts in this thread. It looks like we're now consulting album's `is_downloadable` without checking first if the album is public. I believe that's wrong. `is_downloadable` can be set for public albums only, so its value is only valid in that case. For non-public albums, we should be checking other criteria (either the current user is the owner or the album is shared with them -- in the latter case, we need toconsult the global config `downloadable`).

Reply:
I did apply the fix of 1430.